frontend/dockerfile/parser/parser.go

```diff
@@ -399,30 +399,58 @@ func Parse(rwc io.Reader) (*Result, error) {
 	}, withLocation(handleScannerError(scanner.Err()), currentLine, 0)
 }
 
+// Extracts a heredoc from a possible heredoc regex match
 func heredocFromMatch(match []string) (*Heredoc, error) {
 	if len(match) == 0 {
 		return nil, nil
 	}
 
-	fileDescriptor, _ := strconv.ParseUint(match[1], 10, 0)
+	fd, _ := strconv.ParseUint(match[1], 10, 0)
 	chomp := match[2] == "-"
-	quoteOpen := match[3]
-	name := match[4]
-	quoteClose := match[5]
+	rest := match[3]
+
+	if len(rest) == 0 {
+		return nil, nil
+	}
+
+	shlex := shell.NewLex('\\')
+	shlex.SkipUnsetEnv = true
+
+	// Attempt to parse both the heredoc both with *and* without quotes.
+	// If there are quotes in one but not the other, then we know that some
+	// part of the heredoc word is quoted, so we shouldn't expand the content.
+	shlex.RawQuotes = false
+	words, err := shlex.ProcessWords(rest, []string{})
+	if err != nil {
+		return nil, err
+	}
+	// quick sanity check that rest is a single word
+	if len(words) != 1 {
+		return nil, nil
+	}
+	shlex.RawQuotes = true
+	wordsQuoted, err := shlex.ProcessWords(rest, []string{})
+	if err != nil {
+		return nil, err
+	}
+	if len(wordsQuoted) != len(words) {
+		// some interesting inconsistency in the lexer has happened here...
+		return nil, nil
```

Comment:
Is it correct to return nil in here or should we just set expand false?

Reply:
Hm, ideally I think we *should* panic here as I understand the lexer? It should be impossible for it to be parsed into a different number of words each time? Not quite sure what your personal preference is in cases like this, I'd personally think to panic, but possibly we could just return an error?

Reply:
Panic is too scary. I haven't looked at too deeply how safe `RawQuotes` change is. I think the options in here are either to disable expand or return an error. Ignoring heredoc without an error should be avoided.

Reply:
Have pushed a fix to error slightly more neatly.